inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -2,33 +2,1037 @@
+
+        bitonic_sort_vector(jcp_.bitonic_size, jcp_.axis_dim, tail);
+        if (jcp_.sort_index) {
+            bitonic_sort_vector(jcp_.bitonic_k_size, jcp_.top_k, tail, false);
+        }
+
+        // prc => dst
+        blocked_innermost_store(tail);
+    }
+
+    // bitonic_size: length of the total array for sorting, power of 2
+    //          len: length of the array being sorted
+    //        start: start index of the array being sorted
+    //      sub_len: half of len
+    //    sub_start: start index of the sub array being sorted
+    //    minor_len: half of sub_len
+    //            n: number of valid elements in bitonic sort
+    //            p: pow of 2 number, so that p/2 < n <= p
+    //   empty tail: p-n elements in the rear don't need sorting,
+    inline void bitonic_sort_vector(int p, int n, int elt_num, bool cmp_val = true) {
+        for (int len = 2; len < p; len <<= 1) {
+            for (int start = 0; start < p; start += len) {
+                int sub_len = len >> 1;
+                // empty tail
+                for (int i = sub_len - 1; start + len - i - 1 < n && i >= 0; i--) {
+                    bitonic_swap_vector(start + i, start + len - i - 1, elt_num, cmp_val);
+                }
+                for (; sub_len > 0; sub_len >>= 1) {
+                    for (int sub_start = start; sub_start < start + len; sub_start += sub_len) {
+                        int minor_len = sub_len >> 1;
+                        // empty tail
+                        for (int j = 0; sub_start + j + minor_len < n && j < minor_len; j++) {
+                            bitonic_swap_vector(sub_start + j, sub_start + j + minor_len, elt_num, cmp_val);
+                        }
+                    }
+                }
+            }
+        }
+
+        // last round sort
+        int sub_p = p >> 1;
+        for (int i = sub_p - 1; p - i - 1 < n && i >= 0; i--) {
+            bitonic_swap_vector(i, p - i - 1, elt_num, cmp_val);
+        }
+        for (; sub_p > 0; sub_p >>= 1) {
+            // support partial sort as well as full sort
+            for (int sub_start = 0; (!cmp_val || (cmp_val && sub_start < n)) && sub_start < p;
+                 sub_start += sub_p) {
+                int minor_p = sub_p >> 1;
+                for (int j = 0; sub_start + j + minor_p < n && j < minor_p; j++) {
+                    bitonic_swap_vector(sub_start + j, sub_start + j + minor_p, elt_num, cmp_val);
+                }
+            }
+        }
+    }
+
+    inline void blocked_innermost_load(int elt_num) {
+        for (int i = 0; i < jcp_.axis_dim; i++) {
+            for (int j = 0; j < elt_num; j++) {
+                int offset = i / jcp_.blk_size * jcp_.blk_stride + i % jcp_.blk_size + j * jcp_.blk_size;
+
+                load_scalar(xmm_tmp, ptr[reg_src + offset * jcp_.data_size], jcp_.data_type);
+                store_scalar(ptr[reg_prc + (i * jcp_.sort_stride + j) * jcp_.data_size], xmm_tmp, jcp_.data_type);
```

Comment:
why we can not use load/store emitters here with load/store number of 1?

Reply:
Fixed. Thanks Chenhu!

Reply:
Roll-back to load_scalar and store_scalar for cases of single element load/store, which has better performance. Thanks Chenhu!